test/Microsoft.AspNet.Mvc.Core.Test/Routing/UrlHelperTest.cs

```diff
@@ -851,6 +857,210 @@ public void Action_RouteValueInvalidation_DoesNotAffectActionAndController_Actio
+        {
+            // Arrange
+            var services = GetServices();
+            var router = BuildRouterWithAttributeRoute(services);
+            var urlHelper = CreateUrlHelper("/", router, services);
+
+            // Act
+            var url = urlHelper.Action(
+                action: "Get",
+                controller: "MatchVariable");
+
+            // Assert
+            Assert.Equal("/api/MatchVariable/help", url);
+        }
+
+        private static IRouter BuildRouterWithAttributeRoute(IServiceProvider services)
+        {
+            var routeBuilder = new RouteBuilder
+            {
+                DefaultHandler = new MvcRouteHandler(),
+                ServiceProvider = services
+            };
+
+            routeBuilder.MapRoute("Default", "{controller}/{action}");
+
+            //Adding the attribute route as done in "UseMvc"
+            routeBuilder.Routes.Insert(0, AttributeRouting.CreateAttributeMegaRoute(
+                routeBuilder.DefaultHandler,
+                services));
+
+            return routeBuilder.Build();
+        }
+
+        private static IServiceProvider GetServices()
+        {
+            var invoker = new Mock<IActionInvoker>();
+            invoker
+                .Setup(i => i.InvokeAsync())
+                .Returns(Task.FromResult(true));
+
+            var invokerFactory = new Mock<IActionInvokerFactory>();
+            invokerFactory
+                .Setup(f => f.CreateInvoker(It.IsAny<ActionContext>()))
+                .Returns<ActionContext>((c) =>
+                {
+                    return invoker.Object;
+                });
+
+            var optionsAccessor = new Mock<IOptions<RouteOptions>>();
+            optionsAccessor
+                .SetupGet(o => o.Value)
+                .Returns(new RouteOptions());
+            var loggerFactory = NullLoggerFactory.Instance;
+
+            var serviceMock = new Mock<IServiceProvider>();
+            serviceMock
+                .Setup(s => s.GetService(typeof(IOptions<RouteOptions>)))
+                .Returns(optionsAccessor.Object);
+            serviceMock
+                .Setup(s => s.GetService(typeof(ILoggerFactory)))
+                .Returns(loggerFactory);
+
+            var actionContextAccessor = new ActionContextAccessor()
+            {
+                ActionContext = new ActionContext()
+                {
+                    HttpContext = new DefaultHttpContext()
+                    {
+                        ApplicationServices = serviceMock.Object,
+                        RequestServices = serviceMock.Object,
+                    },
+                    RouteData = new RouteData(),
+                },
+            };
+
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddMvc();
+
+            serviceCollection.AddInstance<ILibraryManager>(CreateLibraryManager());
+            serviceCollection.AddInstance<IActionInvokerFactory>(invokerFactory.Object);
```

Comment:
There are certainly services in the new function which didn't exist in the old one. I actually went back and forth on weather I wanted to merge the two together, if we want to separate them I'm fine with that. Unfortunately everything I added here I ended up needing because even if it doesn't get used by our case, the underlying code will still complain that the service isn't registered. IActionContextAccessor, IActionContextAccessor, IActionSelector, ILoggerFactory, and DiagnosticSource for example are all needed so that EnsureServices in MvcRouteHandler doesn't blow up. I can have a second go at Mocking out more thing without mocking out the behavior we need to try to decrease the number if we want.